### PR TITLE
fix `kosmos2` tests

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

[VLMs]  support attention backends (#37576) actually breaks `kosmos2` as `KosmosTextAttention` is used in the decoder (`Kosmos2TextBlock`) as well as `Kosmos2ImageToTextProjection` (which should attend to all image places).

But without `is_causal`, the `sdpa_attention_forward` will treat it as `causal` due to

```
    if torch.jit.is_tracing() and isinstance(is_causal, torch.Tensor):
        is_causal = is_causal.item()
```

Maybe there is a better way to handle this. But I would not spend too much time but just adding `is_causal` arugment and pass it.

All tests pass on A10 now

### CODE REVIEW

File: tests/models/kosmos2/test_modeling_kosmos2.py [hunk 1]
```diff
@@ -893,15 +893,14 @@ def test_inference_interpolate_pos_encoding(self):
...
-        expected_shape = torch.Size((1, 145, 1024))
+        # (PR 37743 makes `kosmos2` not returning anymore `vision_model_output`)
+        # verify `image_embeds`
+        expected_shape = torch.Size((1, 64, 2048))
 
-        self.assertEqual(outputs.vision_model_output.last_hidden_state.shape, expected_shape)
+        self.assertEqual(outputs.image_embeds.shape, expected_shape)
 
         expected_slice = torch.tensor(
-            [[0.9148, -1.4148, 3.8040], [3.3443, 1.9478, 0.2080], [1.6604, 2.8184, -0.3618]]
```
Comment (ydshieh) [inline tests/models/kosmos2/test_modeling_kosmos2.py:902]: this value was obtained from the buggy modeling code . Now since w fix the code, we need to update the value.

**Review by dcq01 — COMMENTED** [review]
Hi! I'm a grad student working on a research project about using large language models to automate code review. Based on your commit 914b2a9a239cb6f2f946c632bb7ece879a136385 and the changes in tests/models/kosmos2/test_modeling_kosmos2.py, my tool generated this comment:
1. **Error Handling**: The test correctly checks for a `ValueError` when `interpolate_pos_encoding` is set to `False`. This is a good practice to ensure that the model behaves as expected under different configurations.
2. **Verification of New Expected Values**: Ensure that the new expected values in `expected_slice` are correct and reflect the intended output of the model after the changes made in the model or its parameters. If these values are not verified against the model's output, it could lead to false positives in the test results.
3. **Consistency with Model Output**: If the model's behavior has changed (e.g., due to updates in the model architecture or weights), the expected values should be recalibrated accordingly. Ensure that the new expected values are derived from a reliable source or a recent run of the model.
4. **Testing Precision**: The use of `rtol=1e-4` and `atol=1e-4` in `torch.testing.assert_close` is appropriate for comparing floating-point values. However, if the model's output is known to have higher variability, consider adjusting these tolerances to avoid false negatives in the test.
5. **Use of Constants for Tolerances**: Define `RTOL` and `ATOL` constants at the beginning of the test method to improve maintainability.
6. **Memory Usage**: Create the `expected_slice` tensor directly on the target device to save the overhead of moving it later.
7. **Consistent Formatting**: Consider reverting the `torch.testing.assert_close` call to a multi-line format for better readability.

As part of my research, I'm trying to understand how useful these comments are in real-world development. If you have a moment, I'd be super grateful if you could quickly reply to these two yes/no questions:
1. Does this comment provide suggestions from a dimension you hadn’t considered?
2. 2. Do you find this comment helpful?

Thanks a lot for your time and feedback! And sorry again if this message is a bother.

**Review by zucchini-nlp — COMMENTED** [review]
Let's do `is_causal = True if is_decoder else False`, I think that makes sense and removes redundant args passing around

**Comment by ydshieh** [comment]
I love the suggestion, unfortunately there is a super nit issue.

```
class Kosmos2TextBlock

        if config.add_cross_attention:
            self.encoder_attn = KosmosTextAttention(
                config,
                embed_dim=self.embed_dim,
                num_heads=config.attention_heads,
                dropout=config.attention_dropout,
                is_decoder=True,
                add_inner_attn_layernorm=False,
            )
```

At the time back, I added this block (so a decoder could do self-attention as well as cross-attention to an encoder's output, just like the original `Attention is all you need`).

Although, this is never used for Kosmos2 ( microsoft uses fairseq to contain a lot of code in the library, not all the paths are used for a particular model).

So if a user have `config.add_cross_attention`, it will add `add_cross_attention` to a `Kosmos2TextBlock` which is a decoder, but it should still pass `is_causal = False.` 

`Kosmos2ImageToTextProjection` is however has `is_decoder = False`, so would work with your suggestion.

I would say no hub repository of kosmos2 use `config.add_cross_attention=True` however. So do you want me to remove that confusing part of code and go with your suggeestion?

**Comment by zucchini-nlp** [comment]
> I would say no hub repository of kosmos2 use config.add_cross_attention=True however. So do you want me to remove that confusing part of code and go with your suggeestion?

Ah I see, so we have a not-used code path. I think we can remove it with a minor deprecation cycle, which is in line with the whole "unbloating" cycle we have currently. And then we can safely assume that non-decoder uses non-causal mask

**Comment by ydshieh** [comment]
OK, will do that. Thanks
